src/python/WMCore/Services/WMStats/WMStatsPycurl.py

```diff
@@ -0,0 +1,119 @@
+        thisQuery["numOfError"] = row["value"]
+        # add query to the pool
+        paramQueries.append(thisQuery)
+
+    return jobDetailByTasksPycurl(paramQueries, sampleSize, serviceOpts)
+
+
+def jobDetailByTasksPycurl(queries, limit, serviceOpts):
+    """
+    Pycurl-based implementation of WMStatsReader.jobDetailByTasks.
+    In short, for each tuple of errors, it returns "limit" number of documents.
+    :param queries: list of CouchDB query parameters
+    :param limit: number of documents to retrieve from CouchDB
+    :param serviceOpts: dictionary with CouchDB options (url, db name and couchapp name)
+    :return: dictionary with job detail information
+
+    A decoded example of this query would be:
+        scurl "https://cmsweb-test9.cern.ch/couchdb/wmstats/_design/WMStatsErl3/_view/jobsByStatusWorkflow?reduce=false&include_docs=true&startkey=["WORKFLOW_NAME","/WORKFLOW_NAME/TASK_NAME","jobfailed",8006,"SITE_NAME"]&endkey=["WORKFLOW_NAME","/WORKFLOW_NAME/TASK_NAME","jobfailed",8006,"SITE_NAME",{}]&limit=1&stale=update_after"
+    while encoding the url would result in:
+        scurl "https://****.cern.ch/couchdb/wmstats/_design/WMStatsErl3/_view/jobsByStatusWorkflow?reduce=false&include_docs=true&startkey=%5B%22WORKFLOW_NAME%22%2C+%22%2FWORKFLOW_NAME%2FTASK_NAME%22%2C+%22jobfailed%22%2C+8006%2C+%22SITE_NAME%22%5D&endkey=%5B%22WORKFLOW_NAME%22%2C+%22%2FWORKFLOW_NAME%2FTASK_NAME%22%2C+%22jobfailed%22%2C+8006%2C+%22SITE_NAME%22%2C+%7B%7D%5D&limit=1&stale=update_after"
+
+    Example of output is:
+        {"total_rows":4764135,"offset":4056263,"rows":[
+        {"id":"12703ce5-xxx","key":["WORKFLOW_NAME","/WORKFLOW_NAME/TASK_NAME","jobfailed",8006,"SITE_NAME","https://xxx.cern.ch/couchdb/acdcserver","vocms0255.cern.ch",["Fatal Exception"]],
+         "value":{"id":"12703ce5-xxx","rev":"4-xxx"},
+         "doc":{"_id":"12703ce5-xx","_rev":"4-xxx","wmbsid":2709786,"type":"jobsummary", ...
+
+    where (for the examples above):
+        WORKFLOW_NAME is, e.g.: pdmvserv_Run2017G_DoubleMuon_UL2017_MiniAODv2_BParking_230917_124108_9876
+        TASK_NAME is, e.g.: DataProcessing
+        SITE_NAME is, e.g.: T2_US_Wisconsin
+    """
+    uri = f"couchdb/{serviceOpts['dbName']}/_design/{serviceOpts['couchapp']}/_view/jobsByStatusWorkflow"
+    baseUrl = urljoin(serviceOpts['couchURL'], uri)
+
+    encoder = json.JSONEncoder()
+    urlsPool = []
+    for query in queries:
+        options = {'include_docs': encoder.encode(True),
+                   'reduce': encoder.encode(False),
+                   'startkey': encoder.encode(query["startkey"]),
+                   'endkey': encoder.encode(query["endkey"]),
+                   'limit': encoder.encode(limit)}
+        # we cannot encode the 'stale' parameter
+        options.setdefault("stale", "update_after")
+        # encode url data for the GET request
+        thisUrl = f"{baseUrl}?{urlencode(options, doseq=True)}"
+        urlsPool.append(thisUrl)
+
+    jobInfoDoc = {}
```

Comment:
instead of making one, very nested, dictionary, would it be better only to yield individual data-structures and convert this function to a generator?

Reply:
We don't have any benefits with that because data is further manipulated before it is sent back to the end user. In addition to what is done here, there is also this extra method manipulating/modifying it: https://github.com/dmwm/WMCore/pull/11885/files#diff-2f9af48630184a7dfd398437ef3c7ec5026da5d3a3c3813e39b5f5650421b69fR390